newIDE/app/src/AssetStore/AssetSwappingDialog.js

```diff
@@ -0,0 +1,393 @@
+// @flow
+import { t, Trans } from '@lingui/macro';
+import { I18n } from '@lingui/react';
+import * as React from 'react';
+import Dialog from '../UI/Dialog';
+import FlatButton from '../UI/FlatButton';
+import HelpButton from '../UI/HelpButton';
+import { AssetStore, type AssetStoreInterface } from '.';
+import { type ResourceManagementProps } from '../ResourcesList/ResourceSource';
+import { sendAssetAddedToProject } from '../Utils/Analytics/EventSender';
+import RaisedButton from '../UI/RaisedButton';
+import { AssetStoreContext } from './AssetStoreContext';
+import AssetPackInstallDialog from './AssetPackInstallDialog';
+import {
+  installRequiredExtensions,
+  installPublicAsset,
+  checkRequiredExtensionsUpdateForAssets,
+} from './InstallAsset';
+import { isPrivateAsset } from '../Utils/GDevelopServices/Asset';
+import EventsFunctionsExtensionsContext from '../EventsFunctionsExtensionsLoader/EventsFunctionsExtensionsContext';
+import Window from '../Utils/Window';
+import PrivateAssetsAuthorizationContext from './PrivateAssets/PrivateAssetsAuthorizationContext';
+import useAlertDialog from '../UI/Alert/useAlertDialog';
+import { enumerateAssetStoreIds } from './EnumerateAssetStoreIds';
+import { getAssetShortHeadersToDisplay } from './AssetsList';
+import ErrorBoundary from '../UI/ErrorBoundary';
+import LoaderModal from '../UI/LoaderModal';
+import {
+  serializeToJSObject,
+  unserializeFromJSObject,
+} from '../Utils/Serializer';
+import {
+  useFetchAssets,
+  useExtensionUpdateAlertDialog,
+} from './NewObjectDialog';
+
+const isDev = Window.isDev();
+
+const mergeAnimations = (
+  objectsAnimations: Array<{ name: string }>,
+  assetAnimations: Array<{ name: string }>
+) => {
+  const animations = [];
+  // Ensure the object don't loose any animation.
+  for (const objectAnimation of objectsAnimations) {
+    const assetAnimation = assetAnimations.find(
+      assetAnimation => assetAnimation.name === objectAnimation.name
+    ) || {
+      ...assetAnimations[0],
+      name: objectAnimation.name,
+    };
+    animations.push(assetAnimation);
+  }
+  // Add extra animations from the asset.
+  for (const assetAnimation of assetAnimations) {
+    if (
+      !objectsAnimations.some(
+        objectAnimation => objectAnimation.name === assetAnimation.name
+      )
+    ) {
+      animations.push(assetAnimation);
+    }
+  }
+  return animations;
+};
+
+const swapAsset = (
```

Comment:
I think this is worth a few unit tests :)

Reply:
(Mostly for checking the logic, but usage of these untyped `unserializeFromJSObject` is also making it even more of a good idea to cover this)